### ISSUE TITLE
ThisType has inconsistent behavior between arrow functions and normal functions

### ISSUE THREAD

**Issue by Nctdtman** [microsoft/TypeScript#44460]
# Bug Report

### 🔎 Search Terms

ThisType, function, behavior
### 🕗 Version & Regression Information

- This is the behavior in every version I tried
### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.3.2#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwBUQBnDAHgBEAaeAYXhAA8MRVhj4AlcHGYM0jCyoA5jQAUAOmmwRxAFzwoqAJ4BKeAF4AfEtXbt4nAAcMC+AG8AUPHjBF4jTvjUb8MIvoAyQgAssxAQqxiCU2lYAvmqKAG44WMBWVkSk4ta29vCOWrrptvBwGMgw+Hn5SooAjABMAMxuthFuEVRuHpYNdtll+Rj+xJJQnbaFxfh9AZ1NjZFqSSkYaW7A3Z2jJR3ltlBVdVPNrbbtPSsaPbYTA0NbBSBFG5f7M1FWQA)

### 💻 Code

```ts
declare function Test<D, C extends Record<string, (...args: any) => any>>(opts: {
  d: () => D,
  c: C & ThisType<D>
}): void

Test({
  d: () => {
    return {
      a: 123
    }
  },
  c: {
    d() {
      this.a
      return this
    }
  }
})

Test({
  d() {
    return {
      a: 123
    }
  },
  c: {
    d() {
      this.a
      return this
    }
  }
})
```

### 🙁 Actual behavior

In the second use case `this.a` will report an error: `Object is of type 'unknown'`

### 🙂 Expected behavior

Don't report the error.
It seems that it may be a bug when the `Record` type and `ThisType` type meet.

**Comment by RyanCavanaugh** [comment]
See https://github.com/microsoft/TypeScript/issues/33392#issuecomment-531279979

**Comment by Nctdtman** [comment]
> See [#33392 (comment)](https://github.com/microsoft/TypeScript/issues/33392#issuecomment-531279979)

c isn't sub-object. c type is C & ThisType<D>.

And In the first example, the type can be derived, but the second doesn't.

**Comment by dnt1996** [comment]
> > See [#33392 (comment)](https://github.com/microsoft/TypeScript/issues/33392#issuecomment-531279979)
> 
> c isn't sub-object. c type is C & ThisType.
> 
> And In the first example, the type can be derived, but the second doesn't.

the same problem
```typescript
type Options<T> = { foo(): T, bar?: () => number} & ThisType<T>

declare function define<T>(options: Options<T>): void;


define({
    foo() {
        return {
            a: 1
        }
    },

    bar() {
        return this.a
    }
})

```

https://www.typescriptlang.org/play?ts=4.4.0-beta#code/C4TwDgpgBA8mwEsD2A7AzgHgCoD4oF4oBvKAMySQAoBKALiiwBooAjAQwCcB+emgvFAFcAtiwgcAvlABkDABYI0WcBGw4AUOoAmEAMYAbTtFKCUuxKig7SCFKtyUk8ZOnpwL6NXSgA3JAi0Abk1tCBs7SiJ1KBiyCj4o2KSoDghgQQ4UYmjk3LZ6AEYc3IliqAlGTST2DgSypNT0zKhgBTQAOjYy0olqdSA
